[CI] ShardFollowTaskReplicationTests#testFailLeaderReplicaShard failure

Does not reproduce.

```
REPRODUCE WITH: ./gradlew :x-pack:plugin:ccr:test -Dtests.seed=21F92F032FE75068 -Dtests.class=org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests -Dtests.method="testFailLeaderReplicaShard" -Dtests.security.manager=true -Dtests.locale=fr-BE -Dtests.timezone=Indian/Mahe -Dcompiler.java=10 -Druntime.java=10
05:21:58   2> REPRODUCE WITH: ./gradlew :x-pack:plugin:ccr:test -Dtests.seed=21F92F032FE75068 -Dtests.class=org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests -Dtests.method="testFailLeaderReplicaShard" -Dtests.security.manager=true -Dtests.locale=fr-BE -Dtests.timezone=Indian/Mahe -Dcompiler.java=10 -Druntime.java=10
05:21:58   1> [2018-09-06T09:21:58,376][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testFailLeaderReplicaShard] after test
05:21:58 ERROR   11.1s J3 | ShardFollowTaskReplicationTests.testFailLeaderReplicaShard <<< FAILURES!
05:21:58    > Throwable #1: java.lang.AssertionError: 
05:21:58    > Expected: <256>
05:21:58    >      but: was <1>
05:21:58    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
05:21:58    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.assertAllEqual(ESIndexLevelReplicationTestCase.java:398)
```

Build urls:
* https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA=java10,ES_RUNTIME_JAVA=java10,nodes=virtual&&linux/284/console
* https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-darwin-compatibility/1011/console

I failed again: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-windows-compatibility/7/console

I'm looking into it.

```
09:15:07 Suite: org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests
09:15:07   1> [2018-09-18T12:15:03,671][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testChangeHistoryUUID] before test
09:15:07   1> [2018-09-18T12:15:05,402][INFO ][o.e.x.c.a.ShardFollowNodeTask] [testChangeHistoryUUID] [follow_index][0] Started to follow leader shard [leader_index][0], followGlobalCheckPoint=-1, mappingVersion=1
09:15:07   2> sept. 18, 2018 12:15:07 PM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
09:15:07   2> WARNING: Uncaught exception in thread: Thread[Thread-3,5,TGRP-ShardFollowTaskReplicationTests]
09:15:07   2> java.lang.AssertionError: java.util.concurrent.ExecutionException: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
09:15:07   2> 	at __randomizedtesting.SeedInfo.seed([FCF60CAAE5D18178]:0)
09:15:07   1> [2018-09-18T12:15:05,886][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testChangeHistoryUUID] after test
09:15:07   1> [2018-09-18T12:15:05,893][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testSimpleCcrReplication] before test
09:15:07   2> 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:111)
09:15:07   1> [2018-09-18T12:15:06,284][INFO ][o.e.x.c.a.ShardFollowNodeTask] [testSimpleCcrReplication] [follow_index][0] Started to follow leader shard [leader_index][0], followGlobalCheckPoint=-1, mappingVersion=1
09:15:07   1> [2018-09-18T12:15:07,181][INFO ][o.e.x.c.a.ShardFollowNodeTask] [org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests] [follow_index][0] shard follow task has been stopped
09:15:07   1> [2018-09-18T12:15:07,281][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testSimpleCcrReplication] after test
09:15:08   2> 	at java.lang.Thread.run(Thread.java:748)
09:15:08   2> Caused by: java.util.concurrent.ExecutionException: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
09:15:08   2> 	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:265)
09:15:08   2> 	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:252)
09:15:08   1> [2018-09-18T12:15:07,306][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testFailLeaderReplicaShard] before test
09:15:08   2> 	at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:94)
09:15:08   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:227)
09:15:08   1> [2018-09-18T12:15:07,532][INFO ][o.e.x.c.a.ShardFollowNodeTask] [testFailLeaderReplicaShard] [follow_index][0] Started to follow leader shard [leader_index][0], followGlobalCheckPoint=-1, mappingVersion=1
09:15:08   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.index(ESIndexLevelReplicationTestCase.java:210)
09:15:08   1> [2018-09-18T12:15:07,825][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testFailLeaderReplicaShard] after test
09:15:08 ERROR   0.55s J3 | ShardFollowTaskReplicationTests.testFailLeaderReplicaShard <<< FAILURES!
09:15:08   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.appendDocs(ESIndexLevelReplicationTestCase.java:199)
09:15:08    > Throwable #1: java.lang.AssertionError: 
09:15:08   2> 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:108)
09:15:08    > Expected: <256>
09:15:08    >      but: was <2>
09:15:08   2> 	... 1 more
09:15:08    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
09:15:08   2> Caused by: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
09:15:08   2> 	at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:1543)
09:15:08   2> 	at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:1537)
09:15:08   2> 	at org.elasticsearch.index.shard.IndexShard.acquireReplicaOperationPermit(IndexShard.java:2352)
09:15:08   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef.performOn(ESIndexLevelReplicationTestCase.java:564)
09:15:08   2> 	at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplica(ReplicationOperation.java:160)
09:15:08   2> 	at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplicas(ReplicationOperation.java:148)
09:15:08   2> 	at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:119)
09:15:08   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction.execute(ESIndexLevelReplicationTestCase.java:496)
09:15:08   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:226)
09:15:08   2> 	... 4 more
09:15:08   2> REPRODUCE WITH: ./gradlew :x-pack:plugin:ccr:test -Dtests.seed=FCF60CAAE5D18178 -Dtests.class=org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests -Dtests.method="testFailLeaderReplicaShard" -Dtests.security.manager=true -Dtests.locale=lv -Dtests.timezone=W-SU -Dcompiler.java=10 -Druntime.java=8
09:15:08   2> REPRODUCE WITH: ./gradlew :x-pack:plugin:ccr:test -Dtests.seed=FCF60CAAE5D18178 -Dtests.class=org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests -Dtests.method="testFailLeaderReplicaShard" -Dtests.security.manager=true -Dtests.locale=lv -Dtests.timezone=W-SU -Dcompiler.java=10 -Druntime.java=8
09:15:08   2> NOTE: leaving temporary files on disk at: /var/lib/jenkins/workspace/elastic+elasticsearch+6.x+multijob-unix-compatibility/os/opensuse/x-pack/plugin/ccr/build/testrun/test/J3/temp/org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests_FCF60CAAE5D18178-001
09:15:08   2> NOTE: test params are: codec=Asserting(Lucene70): {}, docValues:{}, maxPointsInLeafNode=1844, maxMBSortInHeap=7.5056449030244785, sim=RandomSimilarity(queryNorm=false): {}, locale=lv, timezone=W-SU
09:15:08   2> NOTE: Linux 4.4.104-39-default amd64/Oracle Corporation 1.8.0_181 (64-bit)/cpus=16,threads=1,free=445402808,total=514850816
09:15:08   2> NOTE: All tests run in this JVM: [CcrStatsCollectorTests, ShardChangesResponseTests, ShardFollowTaskReplicationTests]
09:15:08    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.assertAllEqual(ESIndexLevelReplicationTestCase.java:398)
09:15:08    > 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.testFailLeaderReplicaShard(ShardFollowTaskReplicationTests.java:126)
09:15:08    > 	at java.lang.Thread.run(Thread.java:748)Throwable #2: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=49, name=Thread-3, state=RUNNABLE, group=TGRP-ShardFollowTaskReplicationTests]
09:15:08    > Caused by: java.lang.AssertionError: java.util.concurrent.ExecutionException: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
09:15:08    > 	at __randomizedtesting.SeedInfo.seed([FCF60CAAE5D18178]:0)
09:15:08    > 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:111)
09:15:08    > 	at java.lang.Thread.run(Thread.java:748)
09:15:08    > Caused by: java.util.concurrent.ExecutionException: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
09:15:08    > 	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:265)
09:15:08    > 	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:252)
09:15:08    > 	at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:94)
09:15:08    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:227)
09:15:08    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.index(ESIndexLevelReplicationTestCase.java:210)
09:15:08    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.appendDocs(ESIndexLevelReplicationTestCase.java:199)
09:15:08    > 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:108)
09:15:08    > 	... 1 more
09:15:08    > Caused by: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
09:15:08    > 	at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:1543)
09:15:08    > 	at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:1537)
09:15:08    > 	at org.elasticsearch.index.shard.IndexShard.acquireReplicaOperationPermit(IndexShard.java:2352)
09:15:08    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef.performOn(ESIndexLevelReplicationTestCase.java:564)
09:15:08    > 	at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplica(ReplicationOperation.java:160)
09:15:08    > 	at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplicas(ReplicationOperation.java:148)
09:15:08    > 	at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:119)
09:15:08    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction.execute(ESIndexLevelReplicationTestCase.java:496)
09:15:08    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:226)
09:15:08    > 	... 4 more
09:15:08 Completed [17/25] on J3 in 4.26s, 3 tests, 1 error <<< FAILURES!
```

Same again in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+periodic/112/console

```
11:15:59    > Throwable #1: java.lang.AssertionError: 
11:15:59    > Expected: <256>
11:15:59    >      but: was <2>
11:15:59    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
11:15:59    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.assertAllEqual(ESIndexLevelReplicationTestCase.java:412)
11:15:59    > 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.testFailLeaderReplicaShard(ShardFollowTaskReplicationTests.java:134)
11:15:59    > 	at java.lang.Thread.run(Thread.java:748)
11:15:59    > 	Suppressed: java.lang.AssertionError: 
11:15:59    > Expected: <1538565329404L>
11:15:59    >      but: was <1538565329395L>
11:15:59    > 		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
11:15:59    > 		at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.close(ESIndexLevelReplicationTestCase.java:447)
11:15:59    > 		at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.$closeResource(ShardFollowTaskReplicationTests.java:95)
11:15:59    > 		at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.testFailLeaderReplicaShard(ShardFollowTaskReplicationTests.java:139)
11:15:59    > 		... 37 moreThrowable #2: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=46, name=Thread-2, state=RUNNABLE, group=TGRP-ShardFollowTaskReplicationTests]
11:15:59    > Caused by: java.lang.AssertionError: java.util.concurrent.ExecutionException: java.util.NoSuchElementException: No value present
11:15:59    > 	at __randomizedtesting.SeedInfo.seed([957CFC10C20DF8BE]:0)
11:15:59    > 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:119)
11:15:59    > 	at java.lang.Thread.run(Thread.java:748)
11:15:59    > Caused by: java.util.concurrent.ExecutionException: java.util.NoSuchElementException: No value present
11:15:59    > 	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:265)
11:15:59    > 	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:252)
11:15:59    > 	at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:94)
11:15:59    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:241)
11:15:59    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.index(ESIndexLevelReplicationTestCase.java:224)
11:15:59    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.appendDocs(ESIndexLevelReplicationTestCase.java:213)
11:15:59    > 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:116)
11:15:59    > 	... 1 more
11:15:59    > Caused by: java.util.NoSuchElementException: No value present
11:15:59    > 	at java.util.Optional.get(Optional.java:135)
11:15:59    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef.performOn(ESIndexLevelReplicationTestCase.java:590)
11:15:59    > 	at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplica(ReplicationOperation.java:166)
11:15:59    > 	at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplicas(ReplicationOperation.java:153)
11:15:59    > 	at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:124)
11:15:59    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction.execute(ESIndexLevelReplicationTestCase.java:517)
11:15:59    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:240)
11:15:59    > 	... 4 more
```

again in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=amazon/57/consoleFull

```
06:15:23 ERROR   0.87s J3 | ShardFollowTaskReplicationTests.testFailLeaderReplicaShard <<< FAILURES!
06:15:23    > Throwable #1: java.lang.AssertionError: 
06:15:23   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction.execute(ESIndexLevelReplicationTestCase.java:519)
06:15:23    > Expected: <256>
06:15:23   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:241)
06:15:23    >      but: was <4>
06:15:23    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
06:15:23   2> 	... 4 more
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.assertAllEqual(ESIndexLevelReplicationTestCase.java:413)
06:15:23    > 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.testFailLeaderReplicaShard(ShardFollowTaskReplicationTests.java:140)
06:15:23    > 	at java.lang.Thread.run(Thread.java:748)
06:15:23   2> Caused by: [test/uuid][[test][0]] IllegalIndexShardStateException[CurrentState[CLOSED] operation only allowed when shard state is one of [RECOVERING, POST_RECOVERY, STARTED], origin [REPLICA]]
06:15:23    > 	Suppressed: java.lang.AssertionError: 
06:15:23    > Expected: <[DocIdSeqNoAndTerm{id='eqmia2YBF_ihSIXXxJTA seqNo=0 primaryTerm=92}]>
06:15:23   2> 	at org.elasticsearch.index.shard.IndexShard.ensureWriteAllowed(IndexShard.java:1491)
06:15:23    >      but: was <[DocIdSeqNoAndTerm{id='e6mia2YBF_ihSIXXxJT2 seqNo=1 primaryTerm=92}, DocIdSeqNoAndTerm{id='eqmia2YBF_ihSIXXxJTA seqNo=0 primaryTerm=92}, DocIdSeqNoAndTerm{id='fKmia2YBF_ihSIXXxJT7 seqNo=2 primaryTerm=92}, DocIdSeqNoAndTerm{id='famia2YBF_ihSIXXxJT8 seqNo=3 primaryTerm=92}]>
06:15:23   2> 	at org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:695)
06:15:23    > 		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
06:15:23    > 		at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.close(ESIndexLevelReplicationTestCase.java:451)
06:15:23    > 		at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.testFailLeaderReplicaShard(ShardFollowTaskReplicationTests.java:105)
06:15:23   2> 	at org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnReplica(IndexShard.java:686)
06:15:23    > 		... 37 moreThrowable #2: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=51, name=Thread-3, state=RUNNABLE, group=TGRP-ShardFollowTaskReplicationTests]
06:15:23   2> 	at org.elasticsearch.action.bulk.TransportShardBulkAction.performOpOnReplica(TransportShardBulkAction.java:425)
06:15:23    > Caused by: java.lang.AssertionError: java.util.concurrent.ExecutionException: java.lang.UnsupportedOperationException: failing shard [test][0], node[s1], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=CGyfPmgCRg2RTGGZX8Z1-Q], unassigned_info[[reason=ALLOCATION_FAILED], at[2018-10-13T04:15:20.927Z], failed_attempts[1], delayed=false, details[auto generated for test], allocation_status[no_attempt]] isn't supported. failure: failed to perform indexing on replica [test][0], node[s1], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=CGyfPmgCRg2RTGGZX8Z1-Q], unassigned_info[[reason=ALLOCATION_FAILED], at[2018-10-13T04:15:20.927Z], failed_attempts[1], delayed=false, details[auto generated for test], allocation_status[no_attempt]]
06:15:23    > 	at __randomizedtesting.SeedInfo.seed([246DD1D92AF4EE78]:0)
06:15:23   2> 	at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnReplica(TransportShardBulkAction.java:393)
06:15:23    > 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:125)
06:15:23    > 	at java.lang.Thread.run(Thread.java:748)
06:15:23   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase.executeShardBulkOnReplica(ESIndexLevelReplicationTestCase.java:711)
06:15:23    > Caused by: java.util.concurrent.ExecutionException: java.lang.UnsupportedOperationException: failing shard [test][0], node[s1], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=CGyfPmgCRg2RTGGZX8Z1-Q], unassigned_info[[reason=ALLOCATION_FAILED], at[2018-10-13T04:15:20.927Z], failed_attempts[1], delayed=false, details[auto generated for test], allocation_status[no_attempt]] isn't supported. failure: failed to perform indexing on replica [test][0], node[s1], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=CGyfPmgCRg2RTGGZX8Z1-Q], unassigned_info[[reason=ALLOCATION_FAILED], at[2018-10-13T04:15:20.927Z], failed_attempts[1], delayed=false, details[auto generated for test], allocation_status[no_attempt]]
06:15:23    > 	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:265)
06:15:23    > 	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:252)
06:15:23   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase.access$2100(ESIndexLevelReplicationTestCase.java:103)
06:15:23    > 	at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:94)
06:15:23   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$WriteReplicationAction.performOnReplica(ESIndexLevelReplicationTestCase.java:673)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:242)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.index(ESIndexLevelReplicationTestCase.java:225)
06:15:23   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$WriteReplicationAction.performOnReplica(ESIndexLevelReplicationTestCase.java:658)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.appendDocs(ESIndexLevelReplicationTestCase.java:214)
06:15:23    > 	at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:122)
06:15:23    > 	... 1 more
06:15:23   2> 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef$1.onResponse(ESIndexLevelReplicationTestCase.java:601)
06:15:23    > Caused by: java.lang.UnsupportedOperationException: failing shard [test][0], node[s1], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=CGyfPmgCRg2RTGGZX8Z1-Q], unassigned_info[[reason=ALLOCATION_FAILED], at[2018-10-13T04:15:20.927Z], failed_attempts[1], delayed=false, details[auto generated for test], allocation_status[no_attempt]] isn't supported. failure: failed to perform indexing on replica [test][0], node[s1], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=CGyfPmgCRg2RTGGZX8Z1-Q], unassigned_info[[reason=ALLOCATION_FAILED], at[2018-10-13T04:15:20.927Z], failed_attempts[1], delayed=false, details[auto generated for test], allocation_status[no_attempt]]
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef.failShardIfNeeded(ESIndexLevelReplicationTestCase.java:622)
06:15:23   2> 	... 16 more
06:15:23    > 	at org.elasticsearch.action.support.replication.ReplicationOperation$1.onFailure(ReplicationOperation.java:195)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef$1.onResponse(ESIndexLevelReplicationTestCase.java:606)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef$1.onResponse(ESIndexLevelReplicationTestCase.java:597)
06:15:23    > 	at org.elasticsearch.index.shard.IndexShard$3.onResponse(IndexShard.java:2387)
06:15:23    > 	at org.elasticsearch.index.shard.IndexShard$3.onResponse(IndexShard.java:2365)
06:15:23    > 	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:267)
06:15:23    > 	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:234)
06:15:23    > 	at org.elasticsearch.index.shard.IndexShard.acquireReplicaOperationPermit(IndexShard.java:2364)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef.performOn(ESIndexLevelReplicationTestCase.java:593)
06:15:23    > 	at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplica(ReplicationOperation.java:166)
06:15:23    > 	at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplicas(ReplicationOperation.java:153)
06:15:23    > 	at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:124)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction.execute(ESIndexLevelReplicationTestCase.java:519)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:241)
06:15:23    > 	... 4 more
06:15:23    > Caused by: [test/uuid][[test][0]] IllegalIndexShardStateException[CurrentState[CLOSED] operation only allowed when shard state is one of [RECOVERING, POST_RECOVERY, STARTED], origin [REPLICA]]
06:15:23    > 	at org.elasticsearch.index.shard.IndexShard.ensureWriteAllowed(IndexShard.java:1491)
06:15:23    > 	at org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:695)
06:15:23    > 	at org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnReplica(IndexShard.java:686)
06:15:23    > 	at org.elasticsearch.action.bulk.TransportShardBulkAction.performOpOnReplica(TransportShardBulkAction.java:425)
06:15:23    > 	at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnReplica(TransportShardBulkAction.java:393)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase.executeShardBulkOnReplica(ESIndexLevelReplicationTestCase.java:711)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase.access$2100(ESIndexLevelReplicationTestCase.java:103)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$WriteReplicationAction.performOnReplica(ESIndexLevelReplicationTestCase.java:673)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$WriteReplicationAction.performOnReplica(ESIndexLevelReplicationTestCase.java:658)
06:15:23    > 	at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef$1.onResponse(ESIndexLevelReplicationTestCase.java:601)
06:15:23    > 	... 16 more
```

This looks to be still failing:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=ubuntu&&virtual/4/consoleText
 
```
java.lang.AssertionError: java.util.concurrent.ExecutionException: java.util.NoSuchElementException: No value present
```

@dnhatn can you have another look?

I hit this error in a PR build. Admittedly this was on a feature branch but it was recently merged with master in this commit https://github.com/elastic/elasticsearch/commit/d01436de3c55b9c206b89b868c5281a68b06f7d4

The [console output from jenkins](https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-2/255/console) is mangled but separating the lines marked with `1>`, `2>` and those without either I get. 

```
13:17:46   1> [2018-11-23T13:17:34,379][WARN ][o.e.b.JNANatives         ] [[SUITE-ShardFollowTaskReplicationTests-seed#[C6BC6CCDF1C7D887]]] Unable to lock JVM Memory: error=12, reason=Cannot allocate memory
13:17:46   1> [2018-11-23T13:17:34,385][WARN ][o.e.b.JNANatives         ] [[SUITE-ShardFollowTaskReplicationTests-seed#[C6BC6CCDF1C7D887]]] This can result in part of the JVM being swapped out.
13:17:46   1> [2018-11-23T13:17:34,386][WARN ][o.e.b.JNANatives         ] [[SUITE-ShardFollowTaskReplicationTests-seed#[C6BC6CCDF1C7D887]]] Increase RLIMIT_MEMLOCK, soft limit: 65536, hard limit: 65536
13:17:46   1> [2018-11-23T13:17:34,386][WARN ][o.e.b.JNANatives         ] [[SUITE-ShardFollowTaskReplicationTests-seed#[C6BC6CCDF1C7D887]]] These can be adjusted by modifying /etc/security/limits.conf, for example:
13:17:46   1>   # allow user 'jenkins' mlockall
13:17:46   1>   jenkins soft memlock unlimited
13:17:46   1>   jenkins hard memlock unlimited
13:17:46   1> [2018-11-23T13:17:34,387][WARN ][o.e.b.JNANatives         ] [[SUITE-ShardFollowTaskReplicationTests-seed#[C6BC6CCDF1C7D887]]] If you are logged in interactively, you will have to re-login for the new limits to take effect.
13:17:46   1> [2018-11-23T15:17:37,060][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testSimpleCcrReplication] before test
13:17:46   1> [2018-11-23T15:17:42,198][INFO ][o.e.x.c.a.ShardFollowNodeTask] [testSimpleCcrReplication] [follow_index][0] following leader shard [leader_index][0], follower global checkpoint=[-1], mapping version=[1], settings version=[1]
13:17:46   1> [2018-11-23T15:17:43,751][INFO ][o.e.x.c.a.ShardFollowNodeTask] [org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests] [follow_index][0] shard follow task has been stopped
13:17:46   1> [2018-11-23T15:17:44,078][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testSimpleCcrReplication] after test
13:17:46   1> [2018-11-23T15:17:44,095][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testFailLeaderReplicaShard] before test
13:17:46   1> [2018-11-23T15:17:44,355][INFO ][o.e.x.c.a.ShardFollowNodeTask] [testFailLeaderReplicaShard] [follow_index][0] following leader shard [leader_index][0], follower global checkpoint=[-1], mapping version=[1], settings version=[1]
13:17:46   1> [2018-11-23T15:17:44,599][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testFailLeaderReplicaShard] after test
13:17:46   1> [2018-11-23T15:17:44,656][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testChangeFollowerHistoryUUID] before test
13:17:46   1> [2018-11-23T15:17:44,772][INFO ][o.e.x.c.a.ShardFollowNodeTask] [testChangeFollowerHistoryUUID] [follow_index][0] following leader shard [leader_index][0], follower global checkpoint=[-1], mapping version=[1], settings version=[1]
13:17:46   1> [2018-11-23T15:17:44,913][WARN ][o.e.x.c.a.ShardFollowNodeTask] [org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests] shard follow task encounter non-retryable error
13:17:46   1> java.lang.IllegalStateException: unexpected history uuid, expected [G-vFe7hdRDWtzbCyLHry7Q], actual [haj2ZP35Q8u1fZgrLmDtow], shard is likely restored from snapshot or force allocated
13:17:46   1>   at org.elasticsearch.xpack.ccr.action.bulk.TransportBulkShardOperationsAction.shardOperationOnPrimary(TransportBulkShardOperationsAction.java:115) ~[main/:?]
13:17:46   1>   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests$CcrAction.performOnPrimary(ShardFollowTaskReplicationTests.java:520) ~[test/:?]
13:17:46   1>   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests$CcrAction.performOnPrimary(ShardFollowTaskReplicationTests.java:508) ~[test/:?]
13:17:46   1>   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$PrimaryRef.perform(ESIndexLevelReplicationTestCase.java:598) ~[framework-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
13:17:46   1>   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$PrimaryRef.perform(ESIndexLevelReplicationTestCase.java:584) ~[framework-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
13:17:46   1>   at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:102) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
13:17:46   1>   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction.execute(ESIndexLevelReplicationTestCase.java:570) ~[framework-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
13:17:46   1>   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests$3.lambda$innerSendBulkShardOperationsRequest$0(ShardFollowTaskReplicationTests.java:422) ~[test/:?]
13:17:46   1>   at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:627) [elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
13:17:46   1>   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_192]
13:17:46   1>   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_192]
13:17:46   1>   at java.lang.Thread.run(Thread.java:748) [?:1.8.0_192]
13:17:46   1> [2018-11-23T15:17:44,925][INFO ][o.e.x.c.a.ShardFollowNodeTask] [org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests] [follow_index][0] shard follow task has been stopped
13:17:46   1> [2018-11-23T15:17:44,926][INFO ][o.e.x.c.a.ShardFollowNodeTask] [org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests] [follow_index][0] shard follow task has been stopped
13:17:46   1> [2018-11-23T15:17:45,006][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testChangeFollowerHistoryUUID] after test
13:17:46   1> [2018-11-23T15:17:45,012][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testAddNewFollowingReplica] before test
13:17:46   1> [2018-11-23T15:17:45,242][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testAddNewFollowingReplica] after test
13:17:46   1> [2018-11-23T15:17:45,249][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testRetryBulkShardOperations] before test
13:17:46   1> [2018-11-23T15:17:45,512][INFO ][o.e.x.c.a.ShardFollowNodeTask] [testRetryBulkShardOperations] [follow_index][0] following leader shard [leader_index][0], follower global checkpoint=[-1], mapping version=[1], settings version=[1]
13:17:46   1> [2018-11-23T15:17:45,548][INFO ][o.e.i.s.IndexShard       ] [org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests] [test][0] primary-replica resync completed with 0 operations
13:17:46   1> [2018-11-23T15:17:45,679][INFO ][o.e.x.c.a.ShardFollowNodeTask] [org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests] [follow_index][0] shard follow task has been stopped
13:17:46   1> [2018-11-23T15:17:45,700][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testRetryBulkShardOperations] after test
13:17:46   1> [2018-11-23T15:17:45,708][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testChangeLeaderHistoryUUID] before test
13:17:46   1> [2018-11-23T15:17:45,813][INFO ][o.e.x.c.a.ShardFollowNodeTask] [testChangeLeaderHistoryUUID] [follow_index][0] following leader shard [leader_index][0], follower global checkpoint=[-1], mapping version=[1], settings version=[1]
13:17:46   1> [2018-11-23T15:17:45,898][WARN ][o.e.x.c.a.ShardFollowNodeTask] [org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests] shard follow task encounter non-retryable error
13:17:46   1> java.lang.IllegalStateException: unexpected history uuid, expected [g_4zPgZOR_GdaX24aTdYQw], actual [3JLN3KM_TDisa-PcdA1yag]
13:17:46   1>   at org.elasticsearch.xpack.ccr.action.ShardChangesAction.getOperations(ShardChangesAction.java:505) ~[main/:?]
13:17:46   1>   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests$3.lambda$innerSendShardChangesRequest$1(ShardFollowTaskReplicationTests.java:445) ~[test/:?]
13:17:46   1>   at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:627) [elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
13:17:46   1>   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_192]
13:17:46   1>   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_192]
13:17:46   1>   at java.lang.Thread.run(Thread.java:748) [?:1.8.0_192]
13:17:46   1> [2018-11-23T15:17:45,951][INFO ][o.e.x.c.a.ShardFollowTaskReplicationTests] [testChangeLeaderHistoryUUID] after test
```

```
13:17:46   2> nov. 23, 2018 3:17:44 PM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
13:17:46   2> AVERTISSEMENT: Uncaught exception in thread: Thread[Thread-3,5,TGRP-ShardFollowTaskReplicationTests]
13:17:46   2> java.lang.AssertionError: java.util.concurrent.ExecutionException: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
13:17:46   2>   at __randomizedtesting.SeedInfo.seed([C6BC6CCDF1C7D887]:0)
13:17:46   2>   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:130)
13:17:46   2>   at java.lang.Thread.run(Thread.java:748)
13:17:46   2> Caused by: java.util.concurrent.ExecutionException: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
13:17:46   2>   at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:265)
13:17:46   2>   at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:252)
13:17:46   2>   at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:94)
13:17:46   2>   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:245)
13:17:46   2>   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.index(ESIndexLevelReplicationTestCase.java:228)
13:17:46   2>   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.appendDocs(ESIndexLevelReplicationTestCase.java:217)
13:17:46   2>   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:127)
13:17:46   2>   ... 1 more
13:17:46   2> Caused by: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
13:17:46   2>   at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:1526)
13:17:46   2>   at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:1520)
13:17:46   2>   at org.elasticsearch.index.shard.IndexShard.innerAcquireReplicaOperationPermit(IndexShard.java:2402)
13:17:46   2>   at org.elasticsearch.index.shard.IndexShard.acquireReplicaOperationPermit(IndexShard.java:2371)
13:17:46   2>   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef.performOn(ESIndexLevelReplicationTestCase.java:643)
13:17:46   2>   at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplica(ReplicationOperation.java:166)
13:17:46   2>   at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplicas(ReplicationOperation.java:153)
13:17:46   2>   at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:124)
13:17:46   2>   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction.execute(ESIndexLevelReplicationTestCase.java:570)
13:17:46   2>   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:244)
13:17:46   2>   ... 4 more
13:17:46   2> nov. 23, 2018 3:17:44 PM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
13:17:46   2> AVERTISSEMENT: Uncaught exception in thread: Thread[elasticsearch[org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests][generic][T#2],5,TGRP-ShardFollowTaskReplicationTests]
13:17:46   2> org.elasticsearch.common.util.concurrent.EsRejectedExecutionException: rejected execution of java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@3393dfb7 on java.util.concurrent.ScheduledThreadPoolExecutor@4e747682[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
13:17:46   2>   at __randomizedtesting.SeedInfo.seed([C6BC6CCDF1C7D887]:0)
13:17:46   2>   at org.elasticsearch.common.util.concurrent.EsAbortPolicy.rejectedExecution(EsAbortPolicy.java:48)
13:17:46   2>   at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
13:17:46   2>   at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:326)
13:17:46   2>   at java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:533)
13:17:46   2>   at org.elasticsearch.threadpool.ThreadPool.schedule(ThreadPool.java:351)
13:17:46   2>   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$createShardFollowTask$11(ShardFollowTaskReplicationTests.java:384)
13:17:46   2>   at org.elasticsearch.xpack.ccr.action.ShardFollowNodeTask.handleFailure(ShardFollowNodeTask.java:427)
13:17:46   2>   at org.elasticsearch.xpack.ccr.action.ShardFollowNodeTask.lambda$sendShardChangesRequest$4(ShardFollowNodeTask.java:278)
13:17:46   2>   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests$3.lambda$innerSendShardChangesRequest$1(ShardFollowTaskReplicationTests.java:464)
13:17:46   2>   at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:627)
13:17:46   2>   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
13:17:46   2>   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
13:17:46   2>   at java.lang.Thread.run(Thread.java:748)
13:17:46   2> REPRODUCE WITH: ./gradlew :x-pack:plugin:ccr:test -Dtests.seed=C6BC6CCDF1C7D887 -Dtests.class=org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests -Dtests.method="testFailLeaderReplicaShard" -Dtests.security.manager=true -Dtests.locale=fr-BE -Dtests.timezone=Asia/Hebron -Dcompiler.java=11 -Druntime.java=8
13:17:46   2> REPRODUCE WITH: ./gradlew :x-pack:plugin:ccr:test -Dtests.seed=C6BC6CCDF1C7D887 -Dtests.class=org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests -Dtests.method="testFailLeaderReplicaShard" -Dtests.security.manager=true -Dtests.locale=fr-BE -Dtests.timezone=Asia/Hebron -Dcompiler.java=11 -Druntime.java=8
13:17:46   2> REPRODUCE WITH: ./gradlew :x-pack:plugin:ccr:test -Dtests.seed=C6BC6CCDF1C7D887 -Dtests.class=org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests -Dtests.method="testFailLeaderReplicaShard" -Dtests.security.manager=true -Dtests.locale=fr-BE -Dtests.timezone=Asia/Hebron -Dcompiler.java=11 -Druntime.java=8
13:17:46    >      but: was <2>
13:17:46   2> NOTE: leaving temporary files on disk at: /var/lib/jenkins/workspace/elastic+elasticsearch+pull-request-2/x-pack/plugin/ccr/build/testrun/test/J3/temp/org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests_C6BC6CCDF1C7D887-001
13:17:46   2> NOTE: test params are: codec=Asserting(Lucene80): {}, docValues:{}, maxPointsInLeafNode=1965, maxMBSortInHeap=5.650392842402848, sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@35d797ae), locale=fr-BE, timezone=Asia/Hebron
13:17:46   2> NOTE: Linux 3.10.0-862.14.4.el7.x86_64 amd64/Oracle Corporation 1.8.0_192 (64-bit)/cpus=16,threads=1,free=473948104,total=513277952
13:17:46   2> NOTE: All tests run in this JVM: [ShardFollowTaskReplicationTests]
```

```
13:17:46 ERROR   0.54s J3 | ShardFollowTaskReplicationTests.testFailLeaderReplicaShard <<< FAILURES!
13:17:46    > Throwable #1: java.lang.AssertionError:
13:17:46    > Expected: <256>
13:17:46    >   at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
13:17:46    >   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.assertAllEqual(ESIndexLevelReplicationTestCase.java:427)
13:17:46    >   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.testFailLeaderReplicaShard(ShardFollowTaskReplicationTests.java:145)
13:17:46    >   at java.lang.Thread.run(Thread.java:748)Throwable #2: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=35, name=Thread-3, state=RUNNABLE, group=TGRP-ShardFollowTaskReplicationTests]
13:17:46    > Caused by: java.lang.AssertionError: java.util.concurrent.ExecutionException: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
13:17:46    >   at __randomizedtesting.SeedInfo.seed([C6BC6CCDF1C7D887]:0)
13:17:46    >   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:130)
13:17:46    >   at java.lang.Thread.run(Thread.java:748)
13:17:46    > Caused by: java.util.concurrent.ExecutionException: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
13:17:46    >   at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:265)
13:17:46    >   at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:252)
13:17:46    >   at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:94)
13:17:46    >   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:245)
13:17:46    >   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.index(ESIndexLevelReplicationTestCase.java:228)
13:17:46    >   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.appendDocs(ESIndexLevelReplicationTestCase.java:217)
13:17:46    >   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testFailLeaderReplicaShard$2(ShardFollowTaskReplicationTests.java:127)
13:17:46    >   ... 1 more
13:17:46    > Caused by: [test/uuid][[test][0]] IndexShardClosedException[CurrentState[CLOSED] operation only allowed when not closed]
13:17:46    >   at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:1526)
13:17:46    >   at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:1520)
13:17:46    >   at org.elasticsearch.index.shard.IndexShard.innerAcquireReplicaOperationPermit(IndexShard.java:2402)
13:17:46    >   at org.elasticsearch.index.shard.IndexShard.acquireReplicaOperationPermit(IndexShard.java:2371)
13:17:46    >   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction$ReplicasRef.performOn(ESIndexLevelReplicationTestCase.java:643)
13:17:46    >   at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplica(ReplicationOperation.java:166)
13:17:46    >   at org.elasticsearch.action.support.replication.ReplicationOperation.performOnReplicas(ReplicationOperation.java:153)
13:17:46    >   at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:124)
13:17:46    >   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationAction.execute(ESIndexLevelReplicationTestCase.java:570)
13:17:46    >   at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.executeWriteRequest(ESIndexLevelReplicationTestCase.java:244)
13:17:46    >   ... 4 moreThrowable #3: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=32, name=elasticsearch[org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests][generic][T#2], state=RUNNABLE, group=TGRP-ShardFollowTaskReplicationTests]
13:17:46    > Caused by: org.elasticsearch.common.util.concurrent.EsRejectedExecutionException: rejected execution of java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@3393dfb7 on java.util.concurrent.ScheduledThreadPoolExecutor@4e747682[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
13:17:46    >   at __randomizedtesting.SeedInfo.seed([C6BC6CCDF1C7D887]:0)
13:17:46    >   at org.elasticsearch.common.util.concurrent.EsAbortPolicy.rejectedExecution(EsAbortPolicy.java:48)
13:17:46    >   at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
13:17:46    >   at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:326)
13:17:46    >   at java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:533)
13:17:46    >   at org.elasticsearch.threadpool.ThreadPool.schedule(ThreadPool.java:351)
13:17:46    >   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$createShardFollowTask$11(ShardFollowTaskReplicationTests.java:384)
13:17:46    >   at org.elasticsearch.xpack.ccr.action.ShardFollowNodeTask.handleFailure(ShardFollowNodeTask.java:427)
13:17:46    >   at org.elasticsearch.xpack.ccr.action.ShardFollowNodeTask.lambda$sendShardChangesRequest$4(ShardFollowNodeTask.java:278)
13:17:46    >   at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests$3.lambda$innerSendShardChangesRequest$1(ShardFollowTaskReplicationTests.java:464)
13:17:46    >   at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:627)
13:17:46    >   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
13:17:46    >   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
13:17:46    >   at java.lang.Thread.run(Thread.java:748)
```

--- a/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java
+++ b/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java
@@ -106,7 +106,7 @@ public class ShardFollowTaskReplicationTests extends ESIndexLevelReplicationTest
         }
     }
 
-    public void testFailLeaderReplicaShard() throws Exception {
+    public void testAddRemoveShardOnLeader() throws Exception {
         try (ReplicationGroup leaderGroup = createGroup(1 + randomInt(1));
              ReplicationGroup followerGroup = createFollowGroup(randomInt(2))) {
             leaderGroup.startAll();
@@ -120,31 +120,28 @@ public class ShardFollowTaskReplicationTests extends ESIndexLevelReplicationTest
                     leaderSeqNoStats.getMaxSeqNo(),
                     followerSeqNoStats.getGlobalCheckpoint(),
                     followerSeqNoStats.getMaxSeqNo());
-            int docCount = 256;
-            leaderGroup.appendDocs(1);
-            Runnable task = () -> {
-                try {
-                    leaderGroup.appendDocs(docCount - 1);
-                    leaderGroup.syncGlobalCheckpoint();
-                } catch (Exception e) {
-                    throw new AssertionError(e);
+            int batches = between(0, 10);
+            int docCount = 0;
+            for (int i = 0; i < batches; i++) {
+                docCount += leaderGroup.indexDocs(between(1, 5));
+                if (leaderGroup.getReplicas().isEmpty() == false && randomInt(100) < 5) {
+                    IndexShard closingReplica = randomFrom(leaderGroup.getReplicas());
+                    leaderGroup.removeReplica(closingReplica);
+                    closingReplica.close("test", false);
+                    closingReplica.store().close();
+                } else if (leaderGroup.getReplicas().isEmpty() == false && rarely()) {
+                    IndexShard newPrimary = randomFrom(leaderGroup.getReplicas());
+                    leaderGroup.promoteReplicaToPrimary(newPrimary).get();
+                } else if (randomInt(100) < 5) {
+                    leaderGroup.addReplica();
+                    leaderGroup.startReplicas(1);
                 }
-            };
-            Thread thread = new Thread(task);
-            thread.start();
-
-            // Remove and add a new replica
-            IndexShard luckyReplica = randomFrom(leaderGroup.getReplicas());
-            leaderGroup.removeReplica(luckyReplica);
-            luckyReplica.close("stop replica", false);
-            luckyReplica.store().close();
-            leaderGroup.addReplica();
-            leaderGroup.startReplicas(1);
-            thread.join();
-
+                leaderGroup.syncGlobalCheckpoint();
+            }
             leaderGroup.assertAllEqual(docCount);
             assertThat(shardFollowTask.getFailure(), nullValue());
-            assertBusy(() -> followerGroup.assertAllEqual(docCount));
+            int expectedDoc = docCount;
+            assertBusy(() -> followerGroup.assertAllEqual(expectedDoc));
             shardFollowTask.markAsCompleted();
             assertConsistentHistoryBetweenLeaderAndFollower(leaderGroup, followerGroup);
         }

